planner/core/logical_plans.go

```diff
@@ -1199,8 +1199,6 @@ type CTEClass struct {
 	// seedPartPhysicalPlan and recursivePartPhysicalPlan are the physical plans for the seed part and recursive part of this CTE.
 	seedPartPhysicalPlan      PhysicalPlan
 	recursivePartPhysicalPlan PhysicalPlan
-	// cteTask is the physical plan for this CTE, is a wrapper of the PhysicalCTE.
-	cteTask task
```

Comment:
Why  do you remove it？

Reply:
Because the `CTEClass` is shared between different `LogicalCTE`s now and different `LogicalCTE` should generate different `task`. Hence, there is no need to cache the task here.

Reply:
You can refer to https://github.com/pingcap/tidb/pull/25514. We can share the same task without any property.

Reply:
IMHO, the `task` cannot be shared among different `LogicalCTE`s. Because the `PhysicalCTE`s should have different schema.

In the example:
```
mysql> with inv as
    -> (select t1a , t3a, sum(t2c)
    ->     from t1, t2, t3
    ->     where t2a = t1a
    ->         and t2b = t3b
    ->         and t3c = 1998
    ->     group by t1a, t3a)
    -> select inv1.t1a, inv2.t3a
    -> from inv inv1, inv inv2
    -> where inv1.t1a = inv2.t1a
    ->     and inv1.t3a = 4
    ->     and inv2.t3a = 4+1;
```

If the task is shared between `inv1` and `inv2`, they will generate the same task with the same schema in `findBestTask`, which leads to the bug shown in the issue.

Reply:
I test it and it's ok:
``` sql
mysql> with inv as (select t1a , t3a, sum(t2c) from t1, t2, t3 where t2a = t1a and t2b = t3b and t3c = 1998 group by t1a, t3a) select inv1.t1a, inv2.t3a from inv inv1, inv inv2 where inv1.t1a = inv2.t1a and inv1.t3a = 4 and inv2.t3a = 4+1;
Empty set (0.01 sec)

mysql> desc with inv as (select t1a , t3a, sum(t2c) from t1, t2, t3 where t2a = t1a and t2b = t3b and t3c = 1998 group by t1a, t3a) select inv1.t1a, inv2.t3a from inv inv1, inv inv2 where inv1.t1a = inv2.t1a and inv1.t3a = 4 and inv2.t3a = 4+1;
+--------------------------------------------+----------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------+
| id                                         | estRows  | task      | access object | operator info                                                                                                                                  |
+--------------------------------------------+----------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------+
| HashJoin_38                                | 12.49    | root      |               | inner join, equal:[eq(test.t1.t1a, test.t1.t1a)]                                                                                               |
| ├─Selection_42(Build)                      | 12.49    | root      |               | eq(test.t3.t3a, 5), not(isnull(test.t1.t1a))                                                                                                   |
| │ └─CTEFullScan_43                         | 15.61    | root      | CTE:inv2      | data:CTE_0                                                                                                                                     |
| └─Selection_40(Probe)                      | 12.49    | root      |               | eq(test.t3.t3a, 4), not(isnull(test.t1.t1a))                                                                                                   |
|   └─CTEFullScan_41                         | 15.61    | root      | CTE:inv1      | data:CTE_0                                                                                                                                     |
| CTE_0                                      | 15.61    | root      |               | Non-Recursive CTE                                                                                                                              |
| └─Projection_19(Seed Part)                 | 15.61    | root      |               | test.t1.t1a, test.t3.t3a, Column#13                                                                                                            |
|   └─HashAgg_20                             | 15.61    | root      |               | group by:Column#23, Column#24, funcs:sum(Column#20)->Column#13, funcs:firstrow(Column#21)->test.t1.t1a, funcs:firstrow(Column#22)->test.t3.t3a |
|     └─Projection_35                        | 15.61    | root      |               | cast(test.t2.t2c, decimal(32,0) BINARY)->Column#20, test.t1.t1a, test.t3.t3a, test.t1.t1a, test.t3.t3a                                         |
|       └─Projection_21                      | 15.61    | root      |               | test.t1.t1a, test.t2.t2c, test.t3.t3a                                                                                                          |
|         └─HashJoin_23                      | 15.61    | root      |               | inner join, equal:[eq(test.t2.t2a, test.t1.t1a)]                                                                                               |
|           ├─HashJoin_25(Build)             | 12.49    | root      |               | inner join, equal:[eq(test.t3.t3b, test.t2.t2b)]                                                                                               |
|           │ ├─TableReader_28(Build)        | 9.99     | root      |               | data:Selection_27                                                                                                                              |
|           │ │ └─Selection_27               | 9.99     | cop[tikv] |               | eq(test.t3.t3c, 1998), not(isnull(test.t3.t3b))                                                                                                |
|           │ │   └─TableFullScan_26         | 10000.00 | cop[tikv] | table:t3      | keep order:false, stats:pseudo                                                                                                                 |
|           │ └─TableReader_31(Probe)        | 9980.01  | root      |               | data:Selection_30                                                                                                                              |
|           │   └─Selection_30               | 9980.01  | cop[tikv] |               | not(isnull(test.t2.t2a)), not(isnull(test.t2.t2b))                                                                                             |
|           │     └─TableFullScan_29         | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                                                                                 |
|           └─TableReader_34(Probe)          | 9990.00  | root      |               | data:Selection_33                                                                                                                              |
|             └─Selection_33                 | 9990.00  | cop[tikv] |               | not(isnull(test.t1.t1a))                                                                                                                       |
|               └─TableFullScan_32           | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                                                                                 |
+--------------------------------------------+----------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------------------+
21 rows in set (0.00 sec)
```

Here is the diff after applying the change in #25514 
```
diff --git a/planner/core/find_best_task.go b/planner/core/find_best_task.go
index 5dbefee82..220bc1cbd 100644
--- a/planner/core/find_best_task.go
+++ b/planner/core/find_best_task.go
@@ -1972,9 +1972,22 @@ func (ds *DataSource) getOriginalPhysicalIndexScan(prop *property.PhysicalProper
 }

 func (p *LogicalCTE) findBestTask(prop *property.PhysicalProperty, planCounter *PlanCounterTp) (t task, cntPlan int64, err error) {
-       if !prop.IsEmpty() {
+       if !prop.IsEmpty() && !prop.CanAddEnforcer {
                return invalidTask, 1, nil
        }
+       t = p.cte.cteTask
+       defer func() {
+               if err != nil {
+                       return
+               }
+               if !prop.IsEmpty() && prop.CanAddEnforcer {
+                       t = enforceProperty(prop, t, p.basePlan.ctx)
+               }
+       }()
+       if t != nil {
+               // Already built it.
+               return t, 1, nil
+       }
        // The physical plan has been build when derive stats.
        pcte := PhysicalCTE{SeedPlan: p.cte.seedPartPhysicalPlan, RecurPlan: p.cte.recursivePartPhysicalPlan, CTE: p.cte, cteAsName: p.cteAsName}.Init(p.ctx, p.stats)
        pcte.SetSchema(p.schema)
diff --git a/planner/core/logical_plans.go b/planner/core/logical_plans.go
index 7c2a7b8b2..fdd8504e0 100644
--- a/planner/core/logical_plans.go
+++ b/planner/core/logical_plans.go
@@ -1206,6 +1206,7 @@ type CTEClass struct {
        HasLimit bool
        LimitBeg uint64
        LimitEnd uint64
+       cteTask task
 }
```

Reply:
It seems you didn't set the `p.cte.cteTask` with `p.cte.cteTask = t`?

Reply:
Ok, I was wrong.